### ISSUE TITLE
.NET Alloc inaccuracy

### ISSUE THREAD

**Issue by kkokosa** [Microsoft/perfview#553]
During a session with _.NET Alloc_ enabled there are many allocations not registered due to the ```CorProfilerTracer``` unable to get class info. There is nicely described TODO comment but I believe logging is not enough:

```
STDMETHODIMP CorProfilerTracer::ObjectAllocated(ObjectID objectId, ClassID classId) 
{
	EnterCriticalSection(&m_lock);
	// TODO this is probably inefficient, should only call back on classes that are variable sized
	ULONG size = 0;
	m_info->GetObjectSize(objectId, &size);
	ULONG representativeSize = size;

	// We do this for also  the side effect of logging the class  
	ClassInfo* classInfo = GetClassInfo(classId);
	if (classInfo == 0)			// TODO FIX NOW, we should log something.  
		goto Done;
```

Most of the cases of inability to retrieve the ```ClassInfo``` information I examined were failures of ```GetModuleInfo``` call inside ```GetClassInfo```. If something like that happens ```classInfo->ID``` is set to -1 and never tried to be assigned again:

```
ClassInfo* CorProfilerTracer::GetClassInfo(ClassID classId)
{
	// Have I already looked up this class? 
	ClassInfo*& classInfo = m_classInfo[classId];
	if (classInfo == NULL)
		classInfo = new ClassInfo();
	if (classInfo->ID == -1)     // We failed to get info on the class. 
		return NULL;
```

Aside from the fact that the runtime so often does not respond correctly to ```ICorProfilerInfo::GetModuleInfo```, I wonder if it is not better to register such an allocation without detailed information than to simply ignore it?

Possible solution would be as simple as adding some dummy ID assignment in ```GetClassInfo``` method in case of failure:

```
if (classInfo->Name == NULL)
{
	classInfo->Name = new wchar_t[2];
	classInfo->ID = 1;  // added line
	wcscpy_s(classInfo->Name, 2, L"?");
}
```

Assuming testing even the simplest console app:

```
class Program
{
    static void Main(string[] args)
    {
        Console.WriteLine("Hello world!");
        Console.ReadKey();
    }
}
```

Test shows a 588 bytes difference (with and without above simple fix applied) of registered allocations inside Program.Main:

![etwallocs1](https://user-images.githubusercontent.com/11159779/35816218-73c9fbb0-0a9a-11e8-8621-1954f0f9795c.png)

And there are in total 81 ```ETWClrProfiler/ClassIDDefintion``` events with ClassID equal to 1 which means that during even so simple run we would miss object allocations of 81 different classes.

This is even more noticeable on the level of standard library calls:

![gcetw2](https://user-images.githubusercontent.com/11159779/35817138-e33f1a46-0a9c-11e8-9bd6-277aad677cd4.png)

**Comment by vancem** [comment]
I agree this needs to be investigated.   Changing it at least to return a ? when the events don't give the class name (rather than dropping the allocation) is definitely good.    I would like to understand a bit more why we don't have the name, but that is more involved.
